File: python/cg.py
import numpy as np

def conjugate_gradient(A, b, tol=1e-5, max_iter=100):
    """
    Conjugate Gradient solver

    Parameters:
    A (numpy array): matrix
    b (numpy array): right-hand side vector
    tol (float): tolerance for convergence
    max_iter (int): maximum number of iterations

    Returns:
    x (numpy array): solution vector
    """
    n = len(b)
    x = np.zeros(n)
    r = b - np.dot(A, x)
    p = r.copy()
    rho = np.dot(r, r)

    for i in range(max_iter):
        Ap = np.dot(A, p)
        alpha = rho / np.dot(p, Ap)
        x += alpha * p
        r -= alpha * Ap
        rho_new = np.dot(r, r)
        beta = rho_new / rho
        p = r + beta * p
        rho = rho_new

        if np.linalg.norm(r) < tol:
            break

    return x

# Define the matrix A and the right-hand side vector b
A = np.array([[4, -1, 0], [-1, 4, -1], [0, -1, 4]])
b = np.array([1, 2, 3])

# Solve the linear equation system using the Conjugate Gradient solver
x_cg = conjugate_gradient(A, b)

print("Solution using Conjugate Gradient solver:", x_cg)

# Validate the solution using the numpy linalg.solve function
x_numpy = np.linalg.solve(A, b)

print("Solution using numpy linalg.solve function:", x_numpy)

# Check if the solutions are equal
np.testing.assert_allclose(x_cg, x_numpy)
print("The solutions are equal!")

File: matlab/cg.m
% Define the linear equation system
A = [4 -1 0; -1 4 -1; 0 -1 4];  % matrix
b = [1; 2; 3];  % right-hand side vector

% Define the tolerance and maximum number of iterations
tol = 1e-6;
max_iter = 100;

% Initialize the solution vector
x = zeros(size(b));

% Conjugate Gradient solver
r = b - A*x;  % initial residual
p = r;  % initial search direction
rho = dot(r, r);  % initial residual norm
for k = 1:max_iter
    Ap = A*p;  % matrix-vector product
    alpha = rho / dot(p, Ap);  % step size
    x = x + alpha*p;  % update solution
    r = r - alpha*Ap;  % update residual
    rho_new = dot(r, r);  % new residual norm
    if sqrt(rho_new) < tol
        break;
    end
    beta = rho_new / rho;  % update beta
    p = r + beta*p;  % update search direction
    rho = rho_new;  % update residual norm
end

% Validate the solution
residual = norm(b - A*x);
fprintf('Residual norm: %e\n', residual);
if residual < tol
    fprintf('Solution validated!\n');
else
    fprintf('Solution not validated!\n');
end

% Compare with built-in solver (optional)
x_builtin = A\b;
fprintf('Built-in solver solution: \n');
disp(x_builtin);
fprintf('CG solver solution: \n');
disp(x);

